firmware/baseband/proc_btlerx.cpp

```diff
@@ -26,245 +27,476 @@
+    }
+
+    // //Demod the PDU Header
+    uint8_t bit_decision;
+
+    // Jump back down to beginning of PDU header.
+    int sample_idx = symbols_eaten - (8 * num_demod_byte * SAMPLE_PER_SYMBOL);
+
+    uint16_t packet_index = 0;
+
+    for (i = 0; i < num_demod_byte; i++) {
+        rb_buf[packet_index] = 0;
+
+        for (j = 0; j < 8; j++) {
+            I0 = dst_buffer.p[sample_idx].real();
+            Q0 = dst_buffer.p[sample_idx].imag();
+            I1 = dst_buffer.p[sample_idx + 1].real();
+            Q1 = dst_buffer.p[sample_idx + 1].imag();
+
+            bit_decision = (I0 * Q1 - I1 * Q0) > 0 ? 1 : 0;
+            rb_buf[packet_index] = rb_buf[packet_index] | (bit_decision << j);
+
+            sample_idx += SAMPLE_PER_SYMBOL;
+            ;
+        }
+
+        packet_index++;
+    }
+
+    // demod_byte(num_demod_byte, rb_buf);
+
+    scramble_byte(rb_buf, num_demod_byte, scramble_table[channel_number], rb_buf);
+
+    uint8_t pdu_type = (ADV_PDU_TYPE)(rb_buf[0] & 0x0F);
+    // uint8_t tx_add = ((rb_buf[0] & 0x40) != 0);
+    // uint8_t rx_add = ((rb_buf[0] & 0x80) != 0);
+    uint8_t payload_len = (rb_buf[1] & 0x3F);
 
-            if (packet_detected) {
-                skipSamples = 20;
+    // Not valid Advertise Payload.
+    if ((payload_len < 6) || (payload_len > 37)) {
+        return;
+    }
+
+    //--------------Start Payload Parsing--------------------------//
+
+    num_demod_byte = (payload_len + 3);
+    symbols_eaten += 8 * num_demod_byte * SAMPLE_PER_SYMBOL;
+
+    if (symbols_eaten > (int)dst_buffer.count) {
+        return;
+    }
+
+    // sample_idx = symbols_eaten - (8 * num_demod_byte * SAMPLE_PER_SYMBOL);
+
+    for (i = 0; i < num_demod_byte; i++) {
+        rb_buf[packet_index] = 0;
+
+        for (j = 0; j < 8; j++) {
+            I0 = dst_buffer.p[sample_idx].real();
+            Q0 = dst_buffer.p[sample_idx].imag();
+            I1 = dst_buffer.p[sample_idx + 1].real();
+            Q1 = dst_buffer.p[sample_idx + 1].imag();
+
+            bit_decision = (I0 * Q1 - I1 * Q0) > 0 ? 1 : 0;
+            rb_buf[packet_index] = rb_buf[packet_index] | (bit_decision << j);
+
+            sample_idx += SAMPLE_PER_SYMBOL;
+            ;
+        }
+
+        packet_index++;
+    }
+
+    // demod_byte(num_demod_byte, rb_buf + 2);
+
+    scramble_byte(rb_buf + 2, num_demod_byte, scramble_table[channel_number] + 2, rb_buf + 2);
+
+    //--------------Start CRC Checking-----------------------------//
+
+    // Check CRC
+    bool crc_flag = crc_check(rb_buf, payload_len + 2, crc_init_internal);
+    // pkt_count++;
+
+    // This should be the flag that determines if the data should be sent to the application layer.
+    bool sendPacket = false;
+
+    // Checking CRC and excluding Reserved PDU types.
+    if (pdu_type < RESERVED0 && !crc_flag) {
+        if (parse_adv_pdu_payload_byte(rb_buf + 2, payload_len, (ADV_PDU_TYPE)pdu_type, (void*)(&adv_pdu_payload)) == 0) {
+            sendPacket = true;
```

Comment:
Do you need the flag? Seems like it's just sending whenever parse_adv_pdu_payload_byte returns true.

Reply:
Yes we need it as a guard against bad CRC values. The other was filtering out types. I could probably rework this somehow.